### PR TITLE
Convert the activity to take a picture into a fragment and add it to the home page

### PR DESCRIPTION
**Pull request by iliasmerigh**
### In this PR I did the following :

- Converted the activity to take a picture into a fragment and added it to the home page
- Increased the code coverage by testing a non trivial case : Before I wad directly testing the _onCameraActivityResult_ callback by explicitly calling it from the tests so I had to set in non-private; but know I managed to perform the action and wait for the real callback to arrive.
- In consequence I put the callback function in private so the code is more coherent.

### Question for the team :

- The current coverage of the camera fragment is 78%. As you can see on the picture below, the remaining cases are untestable since they depend on the environnement. Should we keep them ? As I said in the comments, checking the camera permissions is not mandatory since the camera activity already checks it so we can directly launch it without checking for permissions. Regarding the SDK version if we check it here, we should also check it everywhere in the code to keep coherence. What do you think about removing these two checks ?

image
image

### CODE REVIEW

File: app/src/androidTest/java/com/github/sdp_begreen/begreen/fragments/CameraFragmentTest.kt [hunk 1]
```diff
@@ -1,41 +1,188 @@
...
+        // Click the add new post button to start the camera Activity
+        onView(withId(R.id.addNewPostBtn)).perform(ViewActions.click())
+
+        // Verify that the camera activity was actually launched
+        Intents.intended(IntentMatchers.hasAction(MediaStore.ACTION_IMAGE_CAPTURE))
+
+        // Check that we resume the previous activity.
+        // When debbuging, we noticed that for this test the value of state was RESUMED, while the state
+        // of the test startShareActivityAfterTakingPhotoWithCameraAndGettingResultOK was CREATED. The state seems to be independent
+        // of the time since we tested at different instants with thread.sleep.
```
Comment (bryanporcello) [inline app/src/androidTest/java/com/github/sdp_begreen/begreen/fragments/CameraFragmentTest.kt:120]: It might be helpful to add additional comments to explain why we care about the state being independent of time and how it affects the test.

File: app/src/androidTest/java/com/github/sdp_begreen/begreen/fragments/CameraFragmentTest.kt
Comment (bryanporcello) [inline app/src/androidTest/java/com/github/sdp_begreen/begreen/fragments/CameraFragmentTest.kt:1]: Overall great comments on the code and good method names that help understand what you are testing. Concerning the tests themselves, it seems that you tested all principal functionalities and consider some edge cases. Well done!

File: app/src/main/java/com/github/sdp_begreen/begreen/fragments/CameraFragment.kt [hunk 2]
```diff
@@ -1,66 +1,128 @@
...
+            onCameraActivityResult(bitmap)
+        }
+    }
+
+    /**
+     * Helper function to start the camera intent, or ask for permission if not granted.
+     */
+    private fun startCameraIntent() {
+
+        // If the camera permission is not granted, ask for it. Otherwise start the camera intent.
```
Comment (bryanporcello) [inline app/src/main/java/com/github/sdp_begreen/begreen/fragments/CameraFragment.kt:85]: Great way to handle camera permission!

File: app/src/main/java/com/github/sdp_begreen/begreen/fragments/CameraFragment.kt
Comment (bryanporcello) [inline app/src/main/java/com/github/sdp_begreen/begreen/fragments/CameraFragment.kt:1]: The code is well-structured, and it is easy to follow the logic and purpose of each function thanks to your comments. Could we add a way to handle cases where the camera activity is canceled or the bitmap is null after taking a picture? Currently, the code doesn't handle these cases, and it may lead to unexpected behavior. What are your thoughts about this?

File: app/src/main/res/layout/fragment_camera.xml
Comment (bryanporcello) [inline app/src/main/res/layout/fragment_camera.xml:1]: All good for me

**Review by bryanporcello — CHANGES_REQUESTED** [review]
Great job Ilias! Your code is following best practices and demonstrates a strong understanding of Android development. Your dedication to studying Android materials is evident in the quality of your code, and it's clear that your hard work has paid off. Your contributions are greatly appreciated and benefit the entire team. Keep up the good work!

File: app/src/androidTest/java/com/github/sdp_begreen/begreen/fragments/CameraFragmentTest.kt (hunk 1 above)
Comment (iliasmerigh) [reply app/src/androidTest/java/com/github/sdp_begreen/begreen/fragments/CameraFragmentTest.kt:120]: Sorry if the comments weren't clear. Actually it was not necessary to mention the time but since I did it I will explain it a bit more.

File: app/src/main/java/com/github/sdp_begreen/begreen/fragments/CameraFragment.kt
Comment (iliasmerigh) [reply app/src/main/java/com/github/sdp_begreen/begreen/fragments/CameraFragment.kt:1]: Actually, if the camera is cancelled it means that the user explicitly pressed the "< back" button of the camera, which will bring him back to the previous activity. Since cancelling actions by going back to previous activity is common, the user understands by himself.
The bitmap is null if-and-only-if the user cancelled the activity, so it's linked to the previous one. In my previous version, the bitmap could be null for some other reasons, but it's not the case anymore and I preferred to keep these tests separated in case we have to make change after.

**Review by bryanporcello — APPROVED** [review]
All good, thanks for the changes! It should have been not easy to take this camera functionality at hand. Bravo for your well-written code !
